### PR TITLE
Initial Dockerized vanilla Spotlight app

### PR DESCRIPTION
**Pull request by bkiahstroud**
**Ticket**: https://github.com/harvard-lts/CURIOSity/issues/1 

# What does this Pull Request do?

This PR sets up a Dockerized, "vanilla" (i.e. no customizations) [Spotlight](https://github.com/projectblacklight/spotlight) app on Rails. 

Versions:
- Ruby v2.7.2
- Rails v6.1.5 
- Spotlight v3.3.0 

# How should this be tested?

Test that app can be spun up locally: 

1. Follow the spin up instructions in the "Running the stack" section of the [README](https://github.com/harvard-lts/CURIOSity/blob/dockerized-vanilla-spotlight-rails-app/README.md) 

# Test coverage

N/A 

# Interested parties

@phil-plencner-hl

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,8 +2,109 @@
...
+
+[Dory](https://github.com/FreedomBen/dory) is required to run this project using Docker. Installation instructions can be found [here](https://github.com/FreedomBen/dory#installation).
+
+Also, you will need to configure Dory to use the `.test` domain; instructions for this can be found [here](https://playbook-staging.notch8.com/en/dev/environment/run-dory-without-password).
+
+#### stack_car
+
+The [stack_car](https://rubygems.org/gems/stack_car) gem is an alternative to docker-compose.
+
+> Stack Car is an opinionated set of tools around Docker and Rails.  It provides convenent methods to start and stop docker-compose, to deploy with rancher and a set of templates to get a new Rails app in to docker as quickly as possible.
```
Comment (bkiahstroud) [inline README.md:30]: @phil-plencner-hl et al -- this is a tool commonly used by our team to interact with Docker, but is not a "hard" requirement. If you and your team would prefer to use docker-compose commands, please let me know and I can update this documentation. If you have any questions about stack_car, feel free to ask!

**Comment by phil-plencner-hl** [comment]
Thanks for your excellent work so far on this!

After reviewing this PR I have some questions and comments.

**1. Spinning up the local environment does not work**

After pulling in this branch, I attempted to spin up the environment using regular docker-compose as well as Stack Car. In both cases the result is essentially the same. The `spotlight` image fails with a permission denied error. Below are examples using both docker-compose and Stack Car

```
docker-compose -f docker-compose.yml up -d --build --force-recreate
WARNING: The Docker Engine you're using is running in swarm mode.

Compose does not use swarm mode to deploy services to multiple nodes in a swarm. All containers will be scheduled on the current node.

To deploy your application across the swarm, use `docker stack deploy`.

Pulling db_migrate (spotlight:)...
ERROR: The image for the service you're trying to recreate has been removed. If you continue, volume data could be lost. Consider backing up your data before continuing.

Continue with the new image? [yN]y
Pulling db_migrate (spotlight:)...
ERROR: pull access denied for spotlight, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
```

```
sc up                
         run  dotenv -f .env.development,.env docker compose up --remove-orphans web from "."
[+] Running 0/3
 ⠿ db_migrate Error                                                                          1.4s
 ⠿ solr Error                                                                                1.4s
 ⠿ web Error                                                                                 1.3s
Error response from daemon: pull access denied for spotlight, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
```

**2. We do not include Postgres or Solr in our Docker stacks** 

Since we have our Solr and Postgres/mySQL servers running outside of our Docker environments, we do not include them in our docker-compose.yml files that get deployed to our Dev / QA / Production environments. If we do need to have a local version of mysql or Solr for individual development we typically put that in a docker-compose-local.yml file to separate it from our normal deployment process. 

If containerizing these work best for your environments, perhaps we could make separate docker YAML files for our individual use and they would both live in the code repository. So a YAML without the Postgres and Solr containers could be in `docker-compose.yml` (for our use) and a YAML with Postgres and Solr containers could be in `docker-compose-scientist.yml` (for your use) so we don't have to do any additional code changes each time there is a PR into our repository.

**3. We would prefer not to use Stack Car or Rancher** 

At this time, we do not want to use Stack Car or Rancher as part of our deployment process. Unless there is a very compelling reason to include it here with the aim towards using it in our other existing Rails applications that we support. 

**4. We do not commit our actual environment variables / .env files to our repositories.** 

Since we usually have different variables set for each of our dev / QA / production environments we typically do not include them in our repositories. Instead, we include an `env-example.txt` file in the repository with the variables names included in it, but not the values themselves. Additionally, we put .env in our `.gitignore` and `.dockerignore` files so we can manage the variables outside of the container. 

A good example to look at would be the Harvard GeoSpatial Library application here:
https://github.com/harvard-huit/Harvard-Geospatial-Library

This example has a `docker-compose-local.yml` file with a mysql container that is not in the regular `docker-compose.yml`, has the `env-example.txt` file and excludes the `.env` in `.gitignore` and `.dockerignore`.

- https://github.com/harvard-huit/Harvard-Geospatial-Library/blob/main/docker-compose-local.yml#L7-L26
- https://github.com/harvard-huit/Harvard-Geospatial-Library/blob/main/env-example.txt
- https://github.com/harvard-huit/Harvard-Geospatial-Library/blob/main/.dockerignore#L1-L5
- https://github.com/harvard-huit/Harvard-Geospatial-Library/blob/main/.gitignore#L24-L25

I think that's everything for now. @dl-maura feel free to add anything I may have missed.

I suggest we meet on Zoom about next steps. I'll work towards setting something up via Slack.

Thank you again. I look forward to discussing it further soon.

**Comment by bkiahstroud** [comment]
The newly added commits fix an issue we discovered with the build. This was not the same issue that you mentioned above: 

> Error response from daemon: pull access denied for spotlight, repository does not exist or may require 'docker login': denied: requested access to the resource is denied

For that error, please try running the following: 

```bash
docker login
docker login registry.gitlab.com
sc up
```

**Comment by phil-plencner-hl** [comment]
I tried following the step above.....but it still does not build the local instance.

```
docker login                                          
Authenticating with existing credentials...
Login Succeeded
```

```
docker login registry.gitlab.com
Authenticating with existing credentials...
Login Succeeded
```

```
sc up                           
         run  dotenv -f .env.development,.env docker compose up --remove-orphans web from "."
[+] Running 0/3
 ⠿ web Error                                                                                     1.3s
 ⠿ db_migrate Error                                                                              1.3s
 ⠿ solr Error                                                                                    1.3s
Error response from daemon: pull access denied for spotlight, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
```

File: README.md (hunk 1 above)
Comment (aprilrieger) [reply README.md:30]: @bkiahstroud The Harvard team voiced that they would like to move forward with docker commands moving forward.

**Comment by bkiahstroud** [comment]
@phil-plencner-hl -- 

My two most recent commits are a first attempt to resolve this feedback you had: 

> We do not include Postgres or Solr in our Docker stacks

I've confirmed that these changes work in our development environment (using the `docker-compose.override.yml` file). When you get the chance, can you please confirm that the default `docker-compose.yml` file will work for you all in your environments?

**Review by phil-plencner-hl — APPROVED** [review]
This works locally for me. I can spin up the local instance of CURIOSity without using Stackcar. The Solr and Postgres containers are in a separate override file that should make it possible for us to manage our internal deployments.

**Pull request merged**